File: Language/C#/Book/tmp/LINQ实战/Chapter09and10.VB/Chapter09and10/Chapter10.vb
Option Infer On

Imports System
Imports System.Collections.Generic
Imports System.Text
Imports NUnit.Framework
Imports System.Linq
Imports System.Xml.Linq
Imports System.Diagnostics
Imports System.Xml.XPath
Imports System.Xml
Imports System.Xml.Xsl
Imports System.IO
Imports System.Runtime.CompilerServices
Imports AmazonCommon

<TestFixture()> Public Class Chapter10

  Private url As String
  Private ns As XNamespace = Amazon.AmazonNS

  Public Sub New()
    Dim parameters = New Dictionary(Of String, String)()
    parameters("Service") = Amazon.ServiceName
    parameters("Version") = Amazon.ServiceVersion
    parameters("Operation") = "TagLookup"
    parameters("TagName") = "dotnet"
    parameters("Count") = "20"
    parameters("ResponseGroup") = "Tags,Small"
    Dim signHelper = New SignedRequestHelper(Amazon.AccessKey, Amazon.SecretKey, "ecs.amazonaws.com")
    url = signHelper.Sign(parameters)
  End Sub


  <Test()> Sub Attributes()

    Dim book As XElement = <book publicationDate="4/14/1978" title="Beauty"/>
    Dim allAttributes As IEnumerable(Of XAttribute) = book.Attributes()
    For Each attribute As XAttribute In allAttributes
      Console.WriteLine(attribute)
    Next

    Dim allPubDateAttributes As IEnumerable(Of XAttribute) = book.Attributes("publicationDate")
    For Each pubDateAttribute As XAttribute In allPubDateAttributes
      Console.WriteLine(CType(pubDateAttribute, DateTime))
    Next

  End Sub

  <Test()> Sub Descendants()
    Dim books As XElement = XElement.Load("categorizedBooks.xml")
    For Each bookElement As XElement In books.Descendants("book")
      Debug.WriteLine(CType(bookElement, String))
    Next
  End Sub



  <Test()> Sub DescendantsVsSelfAndDescendants()
    Dim root As XElement = XElement.Load("categorizedBooks.xml")
    Dim categories As IEnumerable(Of XElement) = root.Descendants("category")

    Console.WriteLine("Descendants")
    For Each categoryElement As XElement In categories
      Debug.WriteLine(" - " + CType(categoryElement.Attribute("name"), String))
    Next

    categories = root.DescendantsAndSelf("category")
    Debug.WriteLine("DescendantsAndSelf")
    For Each categoryElement As XElement In categories
      Debug.WriteLine(" - " + CType(categoryElement.Attribute("name"), String))
    Next
  End Sub

  <Test()> Sub DescendantsWithQueryExpressions()
    Dim root As XElement = XElement.Load("categorizedBooks.xml")
    Dim books = From book In root.Descendants("book") _
     Select CType(book, String)

    For Each book As String In books
      Console.WriteLine(book)
    Next
  End Sub

  <Test()> Sub Ancestors()
    Dim root As XElement = XElement.Load("categorizedBooks.xml")
    Dim dddBook As XElement = root.Descendants("book") _
     .Where(Function(book As XElement) CType(book, String) = "Domain Driven Design") _
     .First()

    ' reverse the order since we want the topmost category first
    Dim ancestors As IEnumerable(Of XElement) = dddBook.Ancestors("category").Reverse()

    ' join each category with a /
    Dim categoryPath As String = String.Join("/", ancestors.Reverse().Select(Function(e As XElement) CType(e.Attribute("name"), String)).ToArray())

    Console.WriteLine(CType(dddBook, String) + " is in the : " + categoryPath + " category.")

  End Sub

  <Test()> Sub Element()
    Dim root As XElement = XElement.Load("categorizedBooks.xml")
    Dim dotNetCategory As XElement = root.Element("category")
    Console.WriteLine(dotNetCategory)
  End Sub

  <Test()> Sub ElementsBeforeSelf()
    Dim root As XElement = XElement.Load("categorizedBooks.xml")
    Dim dddBook As XElement = root.Descendants("book") _
     .Where(Function(book As XElement) CType(book, String) = "Domain Driven Design") _
     .First()

    Dim beforeSelf As IEnumerable(Of XElement) = dddBook.ElementsBeforeSelf()
    For Each element As XElement In beforeSelf
      Console.WriteLine(CType(element, String))
    Next

  End Sub

  <Test()> Sub XPath()
    Dim root As XElement = XElement.Load("categorizedBooks.xml")
    Dim dotNetCategory As XElement = root.XPathSelectElements("//category[@name='.NET']").First()
    Console.WriteLine(dotNetCategory)
  End Sub

  <Test()> Sub XPathSelectElements()
    Dim root As XElement = XElement.Load("categorizedBooks.xml")
    Dim books = From book In root.XPathSelectElements("//book") _
     Select book

    For Each book As XElement In books
      Console.WriteLine(CType(book, String))
    Next
  End Sub

  <Test()> Sub SelectOperator()
    Dim tags As XElement = XElement.Load(url)
    Dim titles = tags.Descendants(ns + "Title").Select(Function(titleElement As XElement) CType(titleElement, String))
    For Each title As String In titles
      Console.WriteLine(title)
    Next
  End Sub

  <Test()> Sub SelectQueryExpression()
    Dim tags As XElement = XElement.Load(url)
    Dim titles = From title In tags.Descendants(ns + "Title") _
     Select CType(title, String)

    For Each title As String In titles
      Console.WriteLine(title)
    Next
  End Sub

  <Test()> Sub Where()
    Dim tags As XElement = XElement.Load(url)

    Dim wpfBooks = From book In tags.Descendants(ns + "Item") _
     Let bookAttributes = book.Element(ns + "ItemAttributes") _
     Let title = CType(bookAttributes.Element(ns + "Title"), String) _
     Where title.Contains("Windows Presentation Foundation") _
     Select title

    For Each title As String In wpfBooks
      Console.WriteLine(title)
    Next

  End Sub

  <Test()> Sub OrderBy()
    Dim tags As XElement = XElement.Load(url)
    Dim titles = From book In tags.Descendants(ns + "Item") _
     Let bookAttributes = book.Element(ns + "ItemAttributes") _
     Let title = CType(bookAttributes.Element(ns + "Title"), String) _
     Order By title _
     Select title

    For Each title As String In titles
      Console.WriteLine(title)
    Next
  End Sub


  <Test()> Sub GroupBy()
    Dim tags As XElement = XElement.Load(url)
    Dim groups = From book In tags.Descendants(ns + "Item") _
     Let bookAttributes = book.Element(ns + "ItemAttributes") _
     Let title = CType(bookAttributes.Element(ns + "Title"), String) _
     Let publisher = CType(bookAttributes.Element(ns + "Manufacturer"), String) _
     Order By publisher, title _
     Group title By Key = publisher Into Group _
     Select Key, Group

    For Each group In groups
      Console.WriteLine(group.Group.Count().ToString & " book(s) published by " & group.Key)
      For Each title As String In group.Group
        Console.WriteLine(" - " + title)
      Next
    Next
  End Sub

  <Test()> Sub Transformation()
    Dim xml As XElement = _
     New XElement("html", _
     New XElement("body", _
     New XElement("h1", "LINQ Books Library"), _
     New XElement("div", _
     New XElement("b", "LINQ in Action"), _
     "\n" + _
     "      By: Fabrice Marguerie, Steve Eichert\n" + _
     "      Published By: Manning\n" + _
     "    " _
     ), _
     New XElement("div", _
     New XElement("b", "AJAX in Action"), _
     "\n" + _
     "      By: Dave Crane\n" + _
     "      Published By: Manning\n" + _
     "    " _
     ), _
     New XElement("div", _
     New XElement("b", "Patterns of Enterprise Application Architecture"), _
     "\n" + _
     "      By: Martin Fowler\n" + _
     "      Published By: APress\n" + _
     "    " _
     ) _
     ) _
    )

    Dim booksXml As XElement = XElement.Load("books.xml")
    Dim books = From book In booksXml.Descendants("book") _
     Select New With { _
     .Title = CType(book.Element("title"), String), _
     .Publisher = CType(book.Element("publisher"), String), _
     .Authors = String.Join(", ", book.Descendants("author").Select(Function(b) CType(b, String)).ToArray()) _
     }

    For Each book In books
      Console.WriteLine(book.Title)
      Console.WriteLine(book.Publisher)
      Console.WriteLine(book.Authors)
    Next

    Dim html As XElement = _
     New XElement("html", _
     New XElement("body", _
     New XElement("h1", "LINQ Books Library"), _
     From book In booksXml.Descendants("book") _
     Select New XElement("div", _
     New XElement("b", CType(book.Element("title"), String), _
     "\n" + _
     "      By: " + String.Join(", ", book.Descendants("author").Select(Function(b) CType(b, String)).ToArray()) + "\n" + _
     "      Published By: " + CType(book.Element("publisher"), String) + "\n" + _
     "    " _
     ) _
     ) _
     ) _
     )

    Console.WriteLine(html)
  End Sub


  <Test()> Sub TransformationWithXSLT()
    Dim xsl As XDocument = <?xml version='1.0' encoding='UTF-8'?>
                           <xsl:stylesheet version='1.0' xmlns:xsl='http://www.w3.org/1999/XSL/Transform'>
                             <xsl:template match='books'>
                               <html>
                                 <title>Book Catalog</title>
                                 <ul>
                                   <xsl:apply-templates select='book'/>
                                 </ul>
                               </html>
                             </xsl:template>
                             <xsl:template match='book'>
                               <li>
                                 <xsl:value-of select='title'/> by <xsl:apply-templates select='author'/>
                               </li>
                             </xsl:template>
                             <xsl:template match='author'>
                               <xsl:if test='position() > 1'>, </xsl:if>
                               <xsl:value-of select='.'/>
                             </xsl:template>
                           </xsl:stylesheet>

    Dim books As XElement = XElement.Load("books.xml")
    Dim output As XDocument = New XDocument()

    Using writer As XmlWriter = output.CreateWriter()
      Dim xslTransformer As XslCompiledTransform = New XslCompiledTransform()
      xslTransformer.Load(XmlReader.Create(New StringReader(xsl.ToString())))
      xslTransformer.Transform(books.CreateReader(), writer)
    End Using

    Console.WriteLine(output)

    Console.WriteLine("With Extension Method")
    Console.WriteLine(XElement.Load("books.xml").XslTransform(xsl.ToString()))
  End Sub


End Class

Module XmlExtensions
  <Extension()> _
  Function XslTransform(ByVal node As XNode, ByVal xsl As String) As XDocument
    Dim output As XDocument = New XDocument()
    Using writer As XmlWriter = output.CreateWriter()
      Dim xslTransformer As XslCompiledTransform = New XslCompiledTransform()
      xslTransformer.Load(XmlReader.Create(New StringReader(xsl)))
      xslTransformer.Transform(node.CreateReader(), writer)
    End Using
    Return output
  End Function
End Module

File: Language/C#/Book/LINQ实战/第09章 LINQ to XML入门/Chapter09and10.CSharp/Chapter09and10/Chapter10.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Xml;
using System.Xml.Linq;
using System.Xml.XPath;
using System.Xml.Xsl;
using AmazonCommon;
using NUnit.Framework;

namespace LinqInAction.Chapter09and10
{
    [TestFixture]
    public class Chapter10
    {
        private string url;
        private XNamespace ns = Amazon.AmazonNS;

        public Chapter10()
        {
            var parameters = new Dictionary<String, String>() {
                { "Service", Amazon.ServiceName},
                { "Version", Amazon.ServiceVersion },
                { "Operation", "TagLookup" },
                { "TagName", "dotnet" },
                { "Count", "20" },
                { "ResponseGroup", "Tags,Small" }
            };
            var signHelper = new SignedRequestHelper(Amazon.AccessKey, Amazon.SecretKey, "ecs.amazonaws.com");
            url = signHelper.Sign(parameters);
        }

        [Test]
        public void Attributes()
        {

            XElement book = XElement.Parse("<book publicationDate=\"4/14/1978\" title=\"Beauty\"/>");
            IEnumerable<XAttribute> allAttributes = book.Attributes();
            foreach (XAttribute attribute in allAttributes)
            {
                Console.WriteLine(attribute);
            }

            IEnumerable<XAttribute> allPubDateAttributes = book.Attributes("publicationDate");
            foreach (XAttribute pubDateAttribute in allPubDateAttributes)
            {
                Console.WriteLine((DateTime)pubDateAttribute);
            }
        }

        [Test]
        public void Descendants()
        {
            XElement books = XElement.Load("categorizedBooks.xml");
            foreach (XElement bookElement in books.Descendants("book"))
            {
                Debug.WriteLine((string)bookElement);
            }
        }

        [Test]
        public void DescendantsVsSelfAndDescendants()
        {
            XElement root = XElement.Load("categorizedBooks.xml");
            IEnumerable<XElement> categories = root.Descendants("category");

            Debug.WriteLine("Descendants");
            foreach (XElement categoryElement in categories)
            {
                Debug.WriteLine(" - " + (string)categoryElement.Attribute("name"));
            }

            categories = root.DescendantsAndSelf("category");
            Debug.WriteLine("DescendantsAndSelf");
            foreach (XElement categoryElement in categories)
            {
                Debug.WriteLine(" - " + (string)categoryElement.Attribute("name"));
            }
        }

        [Test]
        public void DescendantsWithQueryExpressions()
        {
            XElement root = XElement.Load("categorizedBooks.xml");
            var books = from book in root.Descendants("book")
                        select (string)book;

            foreach (string book in books)
            {
                Console.WriteLine(book);
            }
        }

        [Test]
        public void Ancestors()
        {
            XElement root = XElement.Load("categorizedBooks.xml");
            XElement dddBook = root.Descendants("book")
                                   .Where(book => (string)book == "Domain Driven Design")
                                   .First();

            // reverse the order since we want the topmost category first
            IEnumerable<XElement> ancestors = dddBook.Ancestors("category").Reverse();

            // join each category with a /
            string categoryPath = String.Join("/", ancestors.Reverse().Select(e => (string)e.Attribute("name")).ToArray());

            Debug.WriteLine((string)dddBook + " is in the : " + categoryPath + " category.");

        }

        [Test]
        public void Element()
        {
            XElement root = XElement.Load("categorizedBooks.xml");
            XElement dotNetCategory = root.Element("category");
            Console.WriteLine(dotNetCategory);
        }

        [Test]
        public void ElementsBeforeSelf()
        {
            XElement root = XElement.Load("categorizedBooks.xml");
            XElement dddBook = root.Descendants("book")
                                    .Where(book => (string)book == "Domain Driven Design")
                                    .First();

            IEnumerable<XElement> beforeSelf = dddBook.ElementsBeforeSelf();
            foreach (XElement element in beforeSelf)
            {
                Console.WriteLine((string)element);
            }

        }

        [Test]
        public void XPath()
        {
            XElement root = XElement.Load("categorizedBooks.xml");
            XElement dotNetCategory = root.XPathSelectElements("//category[@name='.NET']").First();
            Console.WriteLine(dotNetCategory);
        }

        [Test]
        public void XPathSelectElements()
        {
            XElement root = XElement.Load("categorizedBooks.xml");
            var books = from book in root.XPathSelectElements("//book")
                        select book;

            foreach (XElement book in books)
            {
                Console.WriteLine((string)book);
            }
        }

        [Test]
        public void SelectOperator()
        {
            XElement tags = XElement.Load(url);
            var titles = tags.Descendants(ns + "Title").Select(titleElement => (string)titleElement);
            foreach (string title in titles)
            {
                Console.WriteLine(title);
            }
        }

        [Test]
        public void SelectQueryExpression()
        {
            XElement tags = XElement.Load(url);
            var titles = from title in tags.Descendants(ns + "Title")
                         select (string)title;

            foreach (string title in titles)
            {
                Console.WriteLine(title);
            }
        }

        [Test]
        public void Where()
        {
            XElement tags = XElement.Load(url);

            var wpfBooks = from book in tags.Descendants(ns + "Item")
                           let bookAttributes = book.Element(ns + "ItemAttributes")
                           let title = ((string)bookAttributes.Element(ns + "Title"))
                           where title.Contains("Windows Presentation Foundation")
                           select title;

            foreach (string title in wpfBooks)
            {
                Console.WriteLine(title);
            }

        }

        [Test]
        public void OrderBy()
        {
            XElement tags = XElement.Load(url);
            var titles = from book in tags.Descendants(ns + "Item")
                         let bookAttributes = book.Element(ns + "ItemAttributes")
                         let title = (string)bookAttributes.Element(ns + "Title")
                         orderby title
                         select title;

            foreach (var title in titles)
            {
                Console.WriteLine(title);
            }
        }


        [Test]
        public void GroupBy()
        {
            XElement tags = XElement.Load(url);
            var groups = from book in tags.Descendants(ns + "Item")
                         let bookAttributes = book.Element(ns + "ItemAttributes")
                         let title = (string)bookAttributes.Element(ns + "Title")
                         let publisher = (string)bookAttributes.Element(ns + "Manufacturer")
                         orderby publisher, title
                         group title by publisher;

            foreach (var group in groups)
            {
                Console.WriteLine(group.Count() + " book(s) published by " + group.Key);
                foreach (var title in group)
                {
                    Console.WriteLine(" - " + title);
                }
            }
        }

        [Test]
        public void Transformation()
        {
            XElement xml =
                new XElement("html",
                    new XElement("body",
                            new XElement("h1", "LINQ Books Library"),
                            new XElement("div",
                                    new XElement("b", "LINQ in Action"),
                                    "\n" +
                                    "      By: Fabrice Marguerie, Steve Eichert\n" +
                                    "      Published By: Manning\n" +
                                    "    "
                            ),
                            new XElement("div",
                                    new XElement("b", "AJAX in Action"),
                                    "\n" +
                                    "      By: Dave Crane\n" +
                                    "      Published By: Manning\n" +
                                    "    "
                            ),
                            new XElement("div",
                                    new XElement("b", "Patterns of Enterprise Application Architecture"),
                                    "\n" +
                                    "      By: Martin Fowler\n" +
                                    "      Published By: APress\n" +
                                    "    "
                            )
                    )
            );

            XElement booksXml = XElement.Load("books.xml");
            var books = from book in booksXml.Descendants("book")
                        select new
                        {
                            Title = (string)book.Element("title"),
                            Publisher = (string)book.Element("publisher"),
                            Authors = String.Join(", ", book.Descendants("author").Select(b => (string)b).ToArray())
                        };

            foreach (var book in books)
            {
                Debug.WriteLine(book.Title);
                Debug.WriteLine(book.Publisher);
                Debug.WriteLine(book.Authors);
            }

            XElement html =
                new XElement("html",
                    new XElement("body",
                            new XElement("h1", "LINQ Books Library"),
                            from book in booksXml.Descendants("book")
                            select new XElement("div",
                                new XElement("b", (string)book.Element("title")),
                                "\n" +
                                "      By: " + String.Join(", ", book.Descendants("author").Select(b => (string)b).ToArray()) + "\n" +
                                "      Published By: " + (string)book.Element("publisher") + "\n" +
                                "    "
                            )
                    )
            );

            Debug.WriteLine(html);
        }

        [Test]
        public void TransformationWithXSLT()
        {
            string xsl = @"<?xml version='1.0' encoding='UTF-8' ?>
<xsl:stylesheet version='1.0' xmlns:xsl='http://www.w3.org/1999/XSL/Transform'>
    <xsl:template match='books'>
        <html>
            <title>Book Catalog</title>
            <ul>
                <xsl:apply-templates select='book'/>
            </ul>
        </html>
    </xsl:template>
    <xsl:template match='book'>
        <li>
            <xsl:value-of select='title'/> by <xsl:apply-templates select='author'/>
        </li>
    </xsl:template>
    <xsl:template match='author'>
        <xsl:if test='position() > 1'>, </xsl:if>
        <xsl:value-of select='.'/>
    </xsl:template>
</xsl:stylesheet>";

            XElement books = XElement.Load("books.xml");
            XDocument output = new XDocument();
            using (XmlWriter writer = output.CreateWriter())
            {
                XslCompiledTransform xslTransformer = new XslCompiledTransform();
                xslTransformer.Load(XmlReader.Create(new StringReader(xsl)));
                xslTransformer.Transform(books.CreateReader(), writer);
            }
            //   Console.WriteLine(output);

            Console.WriteLine(XElement.Load("books.xml").XslTransform(xsl));
        }


    }
    public static class XmlExtensions
    {
        public static XDocument XslTransform(this XNode node, string xsl)
        {
            XDocument output = new XDocument();
            using (XmlWriter writer = output.CreateWriter())
            {
                XslCompiledTransform xslTransformer = new XslCompiledTransform();
                xslTransformer.Load(XmlReader.Create(new StringReader(xsl)));
                xslTransformer.Transform(node.CreateReader(), writer);
            }
            return output;
        }
    }
}